src/components/rating/rating.tsx

```diff
@@ -131,100 +179,93 @@ export class Rating
-    const { disabled, intlRating, showChip, scale, count, average } = this;
-
     return (
-      <Fragment>
-        <fieldset
-          class="fieldset"
-          disabled={disabled}
-          onBlur={() => (this.hoverValue = null)}
-          onPointerLeave={() => (this.hoverValue = null)}
-          onTouchEnd={() => (this.hoverValue = null)}
-        >
-          <legend class="visually-hidden">{intlRating}</legend>
-          {this.renderStars()}
-        </fieldset>
-        {(count || average) && showChip ? (
-          <calcite-chip scale={scale} value={count?.toString()}>
-            {!!average && <span class="number--average">{average.toString()}</span>}
-            {!!count && <span class="number--count">({count?.toString()})</span>}
-          </calcite-chip>
-        ) : null}
-        <HiddenFormInputSlot component={this} />
-      </Fragment>
+      <Host
+        onBlur={this.handleRatingFocusLeave}
+        onFocus={this.handleRatingFocusIn}
+        onKeyDown={this.handleHostKeyDown}
+        onPointerOut={this.handleRatingPointerOut}
+        onPointerOver={this.handleRatingPointerOver}
+      >
+        <span class="wrapper">
+          <fieldset class="fieldset" disabled={this.disabled}>
+            <legend class="visually-hidden">{this.intlRating}</legend>
+            {this.starsMap.map(
+              ({
+                average,
+                checked,
+                focused,
+                fraction,
+                hovered,
+                id,
+                idx,
+                partial,
+                selected,
+                value
+              }) => {
+                return (
+                  <label
+                    class={{
+                      star: true,
+                      focused,
+                      selected,
+                      hovered,
+                      average,
+                      partial
+                    }}
+                    htmlFor={id}
+                    onPointerDown={this.handleLabelPointerDown}
+                    onPointerOver={this.handleLabelPointerOver}
+                  >
+                    <input
+                      checked={checked}
+                      class="visually-hidden"
+                      disabled={this.disabled || this.readOnly}
+                      id={id}
+                      name={this.guid}
+                      onChange={this.handleInputChange}
+                      onKeyDown={this.handleInputKeyDown}
+                      ref={(el) => {
+                        this.inputRefs[idx] = el;
+                        return (
+                          (value === 1 || value === this.value) &&
+                          (this.inputFocusRef = el as HTMLInputElement)
+                        );
+                      }}
+                      type="radio"
+                      value={value}
+                    />
+                    <StarIcon icon={selected || average ? "star-f" : "star"} scale={this.scale} />
```

Comment:
If we do use a functional component here, would we move the logic for applying the `icon` into the functional component?